src/types/composite/container.ts

```diff
@@ -533,3 +535,106 @@ export class ContainerType<T extends ObjectLike = ObjectLike> extends CompositeT
     return gindices;
   }
 }
+
+export class ContainerTypeHybrid<T extends ObjectLike = ObjectLike> extends ContainerType<T> {
+  /** Method to allow the Node to merkelize the struct */
+  toFullTree(value: T): Tree {
+    return super.struct_convertToTree(value);
+  }
+
+  /** Overrides to return hybrid structure */
+
+  createTreeBacked(tree: Tree): TreeBacked<T> {
+    const value = new ContainerHybridTreeValue(this, tree);
+    return (new Proxy(value, TreeProxyHandler as unknown) as unknown) as TreeBacked<T>;
+  }
+
+  struct_convertToTree(value: T): Tree {
+    const node = new BranchNodeHybrid(this, value);
+    return new Tree(node);
+  }
+
+  tree_defaultNode(): Node {
+    return new BranchNodeHybrid(this, this.struct_defaultValue());
+  }
+
+  tree_convertToStruct(target: Tree): T {
+    return (target.rootNode as BranchNodeHybrid<T>).value;
+  }
+
+  tree_deserializeFromBytes(data: Uint8Array, start: number, end: number): Tree {
+    const value = this.struct_deserializeFromBytes(data, start, end);
+    const node = new BranchNodeHybrid(this, value);
+    return new Tree(node);
+  }
+
+  tree_serializeToBytes(target: Tree, output: Uint8Array, offset: number): number {
+    const {value} = target.rootNode as BranchNodeHybrid<T>;
+    return this.struct_serializeToBytes(value, output, offset);
+  }
+
+  tree_getProperty(target: Tree, property: PropertyKey): Tree | unknown {
+    const {value} = target.rootNode as BranchNodeHybrid<T>;
+    return value[property as keyof T];
+  }
+
+  tree_setProperty(target: Tree, property: PropertyKey, value: Tree | unknown): boolean {
+    const {value: prevNodeValue} = target.rootNode as BranchNodeHybrid<T>;
+
+    // TODO: Should this check for valid field name? Benchmark the cost
+    const newNodeValue = {...prevNodeValue, [property]: value} as T;
+    const node = new BranchNodeHybrid(this, newNodeValue);
+    target.rootNode = node;
+
+    return true;
+  }
+
+  tree_deleteProperty(target: Tree, prop: PropertyKey): boolean {
+    const chunkIndex = Object.keys(this.fields).findIndex((fieldName) => fieldName === prop);
+    if (chunkIndex === -1) {
+      throw new Error("Invalid container field name");
+    }
+    const fieldType = this.fields[prop as string];
+    return this.tree_setProperty(target, prop, fieldType.struct_defaultValue());
+  }
+
+  *tree_iterateValues(target: Tree): IterableIterator<Tree | unknown> {
+    const {value} = target.rootNode as BranchNodeHybrid<T>;
+    for (const propValue of Object.values(value)) {
+      yield propValue;
+    }
+  }
+
+  *tree_readonlyIterateValues(target: Tree): IterableIterator<Tree | unknown> {
+    return yield* this.tree_iterateValues(target);
+  }
+}
+
+export class BranchNodeHybrid<T> extends Node {
+  constructor(readonly type: ContainerTypeHybrid<T>, readonly value: T) {
+    super();
+  }
+
+  get root(): Uint8Array {
```

Comment:
it looks like `left()` and `right()` expect the value can be changed so it always have to recalculate the tree while this `root()` getter try to cache the root.
Also the `_root` has gone in `BranchNode` due to the latest persistent-merkle-tree, let's use h0 condition instead https://github.com/ChainSafe/persistent-merkle-tree/blob/d9d35f8fc373a0e76307841c8d27e42de252cee9/src/node.ts#L45

Reply:
Updated to use new hashObject. left and right don't cache since that would defeat the purpose of memory efficiency